test/topology_experimental_raft/test_cdc_generation_clearing.py

```diff
@@ -0,0 +1,98 @@
+
+    # Wait until the CDC geneneration publisher publishes the first generation and tries to remove it.
+    await log_file.wait_for("CDC generation publisher fiber has nothing to do. Sleeping.")
+
+    cql = manager.get_cql()
+    await wait_for_cql_and_get_hosts(cql, [server], time.time() + 60)
+    query_gen_ids = SimpleStatement(
+        "SELECT id FROM system.cdc_generations_v3 WHERE key = 'cdc_generations'",
+        consistency_level = ConsistencyLevel.ONE)
+
+    # The first and only generation should not be removed.
+    gen_ids = {r.id for r in await cql.run_async(query_gen_ids)}
+    logger.info(f"Generations after clearing attempt: {gen_ids}")
+    assert len(gen_ids) == 1
+
+
+@pytest.mark.asyncio
+async def test_dependency_on_timestamps(manager: ManagerClient):
+    """Test that CDC generations are not removed from CDC_GENERATIONS_V3 when the difference between their timestamp
+       and the topology coordinator's clock is too small. Then, test that the CDC generation publisher removes
+       the clean-up candidate (together with older generations) if its timestamp is old enough."""
+    logger.info("Bootstrapping first node")
+    servers = [await manager.server_add(cmdline=['--logger-log-level', 'storage_service=trace'])]
+
+    log_file1 = await manager.server_open_log(servers[0].server_id)
+    mark: Optional[int] = None
+
+    query_gen_ids = SimpleStatement(
+        "SELECT id FROM system.cdc_generations_v3 WHERE key = 'cdc_generations'",
+        consistency_level = ConsistencyLevel.ONE)
+
+    async def tried_to_remove_new_gen() -> Optional[tuple[int, set[str]]]:
+        await log_file1.wait_for("CDC generation publisher fiber has nothing to do. Sleeping.", mark)
+        new_mark = await log_file1.mark()
+
+        cql = manager.get_cql()
+        await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+        new_gen_ids = {r.id for r in await cql.run_async(query_gen_ids)}
+
+        return (new_mark, new_gen_ids)
+
+    # The first generation should not be removed.
+    mark, gen_ids = await wait_for(tried_to_remove_new_gen, time.time() + 60)
+    logger.info(f"Generations after first clearing attempt: {gen_ids}")
+    assert len(gen_ids) == 1
+    first_gen_id = next(iter(gen_ids))
+
+    logger.info("Bootstrapping second node")
+    servers += [await manager.server_add()]
+
+    # Both generations should not be removed.
+    mark, gen_ids = await wait_for(tried_to_remove_new_gen, time.time() + 60)
+    logger.info(f"Generations after second clearing attempt: {gen_ids}")
+    assert len(gen_ids) == 2 and first_gen_id in gen_ids
+
+    # We enable this injection to stop timestamps from preventing the clearing of CDC generations.
+    await inject_error_one_shot(manager.api, servers[0].ip_addr, "clean_obsolete_cdc_generations_ignore_ts")
+
+    logger.info("Bootstrapping third node")
+    servers += [await manager.server_add()]
+
+    # The first generation should be removed.
```

Comment:
This one is much more obvious, but could add something like " thanks to the above injection"